### ISSUE TITLE
Certificate errors on five of the twelve blessed links

### ISSUE THREAD

**Issue by sophiajt** [servo/servo#11929]
When I click on any of these:

<img width="796" alt="screen shot 2016-06-29 at 10 53 55 am" src="https://cloud.githubusercontent.com/assets/547158/16457098/47464dd0-3de8-11e6-8c74-077be6a15a44.png">

I get a certificate error:

<img width="508" alt="screen shot 2016-06-29 at 10 56 17 am" src="https://cloud.githubusercontent.com/assets/547158/16457105/4d538648-3de8-11e6-8989-477ceee01a75.png">

Incidentally, telling someone they should feel bad for helping to test your product is... an odd choice.  Perhaps a more helpful message would be appropriate?

**Comment by metajack** [comment]
This seems to be #10550

**Comment by metajack** [comment]
@jonathandturner I don't disagree about the message, but it is poking fun at the site that fails not at the user. It's probably time to be slightly less flippant in our error messages considering it's not just us reading them now.

**Comment by jdm** [comment]
@jonathandturner What is the output of `openssl --help` and `openssl version` for you?

**Comment by asajeffrey** [comment]
IRC chat: http://logs.glob.uno/?c=mozilla%23servo&s=29+Jun+2016&e=29+Jun+2016#c466167

TL;DR: on a machine which doesn't have the problem:
```
$ openssl ciphers -v | grep -i sha256
DHE-RSA-AES256-SHA256   TLSv1.2 Kx=DH       Au=RSA  Enc=AES(256)  Mac=SHA256
DHE-DSS-AES256-SHA256   TLSv1.2 Kx=DH       Au=DSS  Enc=AES(256)  Mac=SHA256
DH-RSA-AES256-SHA256    TLSv1.2 Kx=DH/RSA   Au=DH   Enc=AES(256)  Mac=SHA256
DH-DSS-AES256-SHA256    TLSv1.2 Kx=DH/DSS   Au=DH   Enc=AES(256)  Mac=SHA256
AES256-SHA256           TLSv1.2 Kx=RSA      Au=RSA  Enc=AES(256)  Mac=SHA256
ECDHE-RSA-AES128-GCM-SHA256 TLSv1.2 Kx=ECDH     Au=RSA  Enc=AESGCM(128) Mac=AEAD
ECDHE-ECDSA-AES128-GCM-SHA256 TLSv1.2 Kx=ECDH     Au=ECDSA Enc=AESGCM(128) Mac=AEAD
ECDHE-RSA-AES128-SHA256 TLSv1.2 Kx=ECDH     Au=RSA  Enc=AES(128)  Mac=SHA256
ECDHE-ECDSA-AES128-SHA256 TLSv1.2 Kx=ECDH     Au=ECDSA Enc=AES(128)  Mac=SHA256
DH-DSS-AES128-GCM-SHA256 TLSv1.2 Kx=DH/DSS   Au=DH   Enc=AESGCM(128) Mac=AEAD
DHE-DSS-AES128-GCM-SHA256 TLSv1.2 Kx=DH       Au=DSS  Enc=AESGCM(128) Mac=AEAD
DH-RSA-AES128-GCM-SHA256 TLSv1.2 Kx=DH/RSA   Au=DH   Enc=AESGCM(128) Mac=AEAD
DHE-RSA-AES128-GCM-SHA256 TLSv1.2 Kx=DH       Au=RSA  Enc=AESGCM(128) Mac=AEAD
DHE-RSA-AES128-SHA256   TLSv1.2 Kx=DH       Au=RSA  Enc=AES(128)  Mac=SHA256
DHE-DSS-AES128-SHA256   TLSv1.2 Kx=DH       Au=DSS  Enc=AES(128)  Mac=SHA256
DH-RSA-AES128-SHA256    TLSv1.2 Kx=DH/RSA   Au=DH   Enc=AES(128)  Mac=SHA256
DH-DSS-AES128-SHA256    TLSv1.2 Kx=DH/DSS   Au=DH   Enc=AES(128)  Mac=SHA256
ECDH-RSA-AES128-GCM-SHA256 TLSv1.2 Kx=ECDH/RSA Au=ECDH Enc=AESGCM(128) Mac=AEAD
ECDH-ECDSA-AES128-GCM-SHA256 TLSv1.2 Kx=ECDH/ECDSA Au=ECDH Enc=AESGCM(128) Mac=AEAD
ECDH-RSA-AES128-SHA256  TLSv1.2 Kx=ECDH/RSA Au=ECDH Enc=AES(128)  Mac=SHA256
ECDH-ECDSA-AES128-SHA256 TLSv1.2 Kx=ECDH/ECDSA Au=ECDH Enc=AES(128)  Mac=SHA256
AES128-GCM-SHA256       TLSv1.2 Kx=RSA      Au=RSA  Enc=AESGCM(128) Mac=AEAD
AES128-SHA256           TLSv1.2 Kx=RSA      Au=RSA  Enc=AES(128)  Mac=SHA256
```

**Comment by jonathandturner** [comment]
@jdm 

```
jturner-23759:servo jturner$ openssl --help
openssl:Error: '--help' is an invalid command.

Standard commands
asn1parse      ca             ciphers        crl            crl2pkcs7      
dgst           dh             dhparam        dsa            dsaparam       
ec             ecparam        enc            engine         errstr         
gendh          gendsa         genrsa         nseq           ocsp           
passwd         pkcs12         pkcs7          pkcs8          prime          
rand           req            rsa            rsautl         s_client       
s_server       s_time         sess_id        smime          speed          
spkac          verify         version        x509           

Message Digest commands (see the `dgst' command for more details)
md2            md4            md5            mdc2           rmd160         
sha            sha1           

Cipher commands (see the `enc' command for more details)
aes-128-cbc    aes-128-ecb    aes-192-cbc    aes-192-ecb    aes-256-cbc    
aes-256-ecb    base64         bf             bf-cbc         bf-cfb         
bf-ecb         bf-ofb         cast           cast-cbc       cast5-cbc      
cast5-cfb      cast5-ecb      cast5-ofb      des            des-cbc        
des-cfb        des-ecb        des-ede        des-ede-cbc    des-ede-cfb    
des-ede-ofb    des-ede3       des-ede3-cbc   des-ede3-cfb   des-ede3-ofb   
des-ofb        des3           desx           rc2            rc2-40-cbc     
rc2-64-cbc     rc2-cbc        rc2-cfb        rc2-ecb        rc2-ofb        
rc4            rc4-40         seed           seed-cbc       seed-cfb       
seed-ecb       seed-ofb       
```
```
jturner-23759:servo jturner$ openssl version
OpenSSL 0.9.8zh 14 Jan 2016
```

**Comment by jdm** [comment]
@jonathandturner Could you run `openssl ciphers -v | grep -i sha256` to confirm the theory that your openssl doesn't support sha256?

**Comment by jonathandturner** [comment]
@jdm - I don't get anything when I run the command

**Comment by asajeffrey** [comment]
@jdm @jonathandturner it looks quite likely to be sha256 that's causing the problem.

**Comment by frewsxcv** [comment]
I get the Zoidberg error page and my OpenSSL has SHA256 stuffs:

```
$ openssl ciphers -v | grep -i sha256
DHE-RSA-AES256-SHA256   TLSv1.2 Kx=DH       Au=RSA  Enc=AES(256)  Mac=SHA256
DHE-DSS-AES256-SHA256   TLSv1.2 Kx=DH       Au=DSS  Enc=AES(256)  Mac=SHA256
DH-RSA-AES256-SHA256    TLSv1.2 Kx=DH/RSA   Au=DH   Enc=AES(256)  Mac=SHA256
DH-DSS-AES256-SHA256    TLSv1.2 Kx=DH/DSS   Au=DH   Enc=AES(256)  Mac=SHA256
AES256-SHA256           TLSv1.2 Kx=RSA      Au=RSA  Enc=AES(256)  Mac=SHA256
ECDHE-RSA-AES128-GCM-SHA256 TLSv1.2 Kx=ECDH     Au=RSA  Enc=AESGCM(128) Mac=AEAD
ECDHE-ECDSA-AES128-GCM-SHA256 TLSv1.2 Kx=ECDH     Au=ECDSA Enc=AESGCM(128) Mac=AEAD
ECDHE-RSA-AES128-SHA256 TLSv1.2 Kx=ECDH     Au=RSA  Enc=AES(128)  Mac=SHA256
ECDHE-ECDSA-AES128-SHA256 TLSv1.2 Kx=ECDH     Au=ECDSA Enc=AES(128)  Mac=SHA256
DH-DSS-AES128-GCM-SHA256 TLSv1.2 Kx=DH/DSS   Au=DH   Enc=AESGCM(128) Mac=AEAD
DHE-DSS-AES128-GCM-SHA256 TLSv1.2 Kx=DH       Au=DSS  Enc=AESGCM(128) Mac=AEAD
DH-RSA-AES128-GCM-SHA256 TLSv1.2 Kx=DH/RSA   Au=DH   Enc=AESGCM(128) Mac=AEAD
DHE-RSA-AES128-GCM-SHA256 TLSv1.2 Kx=DH       Au=RSA  Enc=AESGCM(128) Mac=AEAD
DHE-RSA-AES128-SHA256   TLSv1.2 Kx=DH       Au=RSA  Enc=AES(128)  Mac=SHA256
DHE-DSS-AES128-SHA256   TLSv1.2 Kx=DH       Au=DSS  Enc=AES(128)  Mac=SHA256
DH-RSA-AES128-SHA256    TLSv1.2 Kx=DH/RSA   Au=DH   Enc=AES(128)  Mac=SHA256
DH-DSS-AES128-SHA256    TLSv1.2 Kx=DH/DSS   Au=DH   Enc=AES(128)  Mac=SHA256
ECDH-RSA-AES128-GCM-SHA256 TLSv1.2 Kx=ECDH/RSA Au=ECDH Enc=AESGCM(128) Mac=AEAD
ECDH-ECDSA-AES128-GCM-SHA256 TLSv1.2 Kx=ECDH/ECDSA Au=ECDH Enc=AESGCM(128) Mac=AEAD
ECDH-RSA-AES128-SHA256  TLSv1.2 Kx=ECDH/RSA Au=ECDH Enc=AES(128)  Mac=SHA256
ECDH-ECDSA-AES128-SHA256 TLSv1.2 Kx=ECDH/ECDSA Au=ECDH Enc=AES(128)  Mac=SHA256
AES128-GCM-SHA256       TLSv1.2 Kx=RSA      Au=RSA  Enc=AESGCM(128) Mac=AEAD
AES128-SHA256           TLSv1.2 Kx=RSA      Au=RSA  Enc=AES(128)  Mac=SHA256
```

On a side note, https://servo.org supports ciphersuites that use SHA1, not jsut SHA256:

TLS_ECDHE_ECDSA_WITH_AES_128_CBC_SHA

**Comment by asajeffrey** [comment]
SHA256 compatibility: https://support.globalsign.com/customer/portal/articles/1499561-sha-256-compatibility

In particular, SHA256 is only supported in OS X 10.5 and above. @jonathandturner are you on 
OS X by any chance?

**Comment by asajeffrey** [comment]
@frewsxcv your openssl may be picking up a different version of the library than servo :(

**Comment by jonathandturner** [comment]
@asajeffrey - I'm on OS X 10.11.5

**Comment by jdm** [comment]
#11932 proposes a more useful SSL error page; unfortunately as long as we use the stock system OpenSSL that's available, we may end up in cases like this one where OpenSSL 0.9.8 simply cannot connect to some sites (see the simulated handshake list at https://dev.ssllabs.com/ssltest/analyze.html?d=servo.org&s=104.31.74.227).

**Comment by asajeffrey** [comment]
I think #11932 is about as good as we'll get right now, so I'll remove the P1 label.

**Comment by frewsxcv** [comment]
Something that's also interesting: I can only get the TLS errors when using release mode. The pages load fine for me in debug mode.

**Comment by frewsxcv** [comment]
Using [this Servo patch](https://gist.github.com/frewsxcv/8c60bf853dcd6b3f60c74512051377f1), I was able to switch the [TLS backend for Hyper to be Security Framework](https://github.com/hyperium/hyper/pull/762) (ignore some of the changes in there about the `connection` field in that struct getting removed. This remedied the security issues I encountered above with OpenSSL.

**Comment by atouchet** [comment]
Is this issue still relevant?

**Comment by frewsxcv** [comment]
Is there another issue tracking improvements to TLS errors?

**Comment by lumiscosity** [comment]
According to https://github.com/servo/servo/pull/30025 we switched from OpenSSL to rustls. The bad cert page has also recieved more updates over time. I'm also not sure what that landing page is? Once again, is this still relevant?

**Comment by mrobinson** [comment]
Unclear if this is still relevant, so I'll close it and we can open a new issue if we find new certificate errors.
